mne_bids/write.py

```diff
@@ -2158,3 +2163,387 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+        bids_path.datatype == 'meg' and
+        bids_path.suffix == 'meg' and
+        bids_path.acquisition == 'calibration' and
+        bids_path.extension == '.dat'
+    )
+    return is_finecal_path
+
+
+@verbose
+def anonymize_dataset(bids_root_in, bids_root_out, daysback='auto',
+                      subject_mapping='auto', datatypes=None,
+                      random_state=None, verbose=None):
+    """Anonymize a BIDS dataset.
+
+    Parameters
+    ----------
+    bids_root_in : path-like
+        The root directory of the input BIDS dataset.
+    bids_root_out : path-like
+        The directory to place the anonymized dataset into.
+    daysback : int | 'auto'
+        Number of days by which to move back the recording date in time. If
+        ``'auto'``, tries to randomly pick a suitable number.
+    subject_mapping : dict | callable | 'auto' | None
+        How to anonymize subject IDs. If a dictionary, maps the original IDs
+        (keys) to the anonymized IDs (values). If a function, must be one that
+        accepts the original IDs as a list of strings and returns a dictionary
+        with original IDs as keys and anonymized IDs as values. If ``'auto'``,
+        automatically produces a mapping (unique, randomly picked 3-digit IDs)
+        and prints it on the screen. If ``None``, subject IDs are not changed.
+    datatypes : list of str | str | None
+        Which data type to anonymize. If can be ``meg``, ``eeg``, ``ieeg``, or
+        ``anat``. Multiple data types may be passed as a collection of strings.
+        If ``None``, try to anonymize the entire input dataset.
+    %(random_state)s
+        The RNG will be used to derive ``daysback`` and ``subject_mapping`` if
+        they are ``'auto'``.
+    %(verbose)s
+    """
+    from mne_bids import update_sidecar_json  # avoid circular import
+
+    bids_root_in = Path(bids_root_in).expanduser()
+    bids_root_out = Path(bids_root_out).expanduser()
+    rng = np.random.default_rng(seed=random_state)
+
+    if not bids_root_in.is_dir():
+        raise FileNotFoundError(
+            f'The specified input directory does not exist: {bids_root_in}'
+        )
+
+    if bids_root_in == bids_root_out:
+        raise ValueError('Input and output directory must differ')
+
+    if bids_root_out.exists():
+        raise FileExistsError(
+            f'The specified output directory already exists. Please remove '
+            f'it to perform anonymization: {bids_root_out}'
+        )
+
+    if not isinstance(subject_mapping, dict):
+        participants_tsv = _from_tsv(bids_root_in / 'participants.tsv')
+        participants_in = [
+            participant.replace('sub-', '')
+            for participant in participants_tsv['participant_id']
+        ]
+
+        if subject_mapping == 'auto':
+            participants_out = rng.choice(
+                a=range(100, 1000),
+                replace=False,
+                size=len(participants_in)
+            )
```

Comment:
I think it's a bit too convoluted. I would generate subject numbers from 1 to n_subjects but use a random permutation. Doing so makes is easy to know how many subjects you have etc. Having a subject number of 332 and you have only one subject suggests it's part of a bigger dataset.

Reply:
to then illustrate this in the example just write first data subjects that initials or subject names and you will see that it will get converted to 01 etc...

Reply:
I understand what you're saying but I'm not 100% convinced yet. What I did here is that I basically implemented how we used to produce pseudonymous IDs in my old lab. I always thought 3-digit IDs were a good idea because they allow for lots of "randomness".

I'd be interested to hear what @sappelhoff and @cbrnr think about this?

Also CC @SophieHerbst

Reply:
I agree with @agramfort that assigning numbers from 1 to N (prepended with as many zeros as necessary) is fine as long as the original subject order is randomly permuted.

Reply:
If N < 10, should we prepend a 0 too?

Reply:
agreed with Alex and Clemens as well :+1: 

> If N < 10, should we prepend a 0 too?

:+1:

Reply:
I wouldn't append a 0 if this is not necessary - why would you do that? Data collection is already done, and if there are N < 10 subjects then I'd call them 1...9.

Reply:
okay, fair - I misread that "N < 10". Right, if there are no numbers beyond 9, I also see no need for zero-padding ... ordering  will also work without that in that case.

Reply:
But we e.g. also always zero-pad runs < 10 … wouldn't that be inconsistent?

Reply:
yeah .. why do we do that even? :thinking: 

https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#participant-names-and-other-labels

> In <index>es we RECOMMEND using zero padding (for example, 01 instead of 1 **if you have more than nine subjects**) to make alphabetical sorting more intuitive. 

**emphasis by me**

Reply:
ok will not pad then  for N < 10 :)